Mathlib/Algebra/Category/Grp/Colimits.lean

```diff
@@ -42,220 +33,181 @@ then taking the quotient by the abelian group laws within each abelian group,
+  dsimp [desc, ι]
+  erw [QuotientAddGroup.lift_mk']
+  simp
 
 @[simp]
-theorem quot_add (x y) :
-    (by exact Quot.mk Setoid.r (add x y) : ColimitType.{w} F) =
-      -- Porting note: force Lean to treat `ColimitType F` no as `Quot _`
-      (by exact Quot.mk Setoid.r x) + (by exact Quot.mk Setoid.r y) :=
-  rfl
-
-/-- The bundled abelian group giving the colimit of a diagram. -/
-def colimit : AddCommGrp :=
-  AddCommGrp.of (ColimitType.{w} F)
-
-/-- The function from a given abelian group in the diagram to the colimit abelian group. -/
-def coconeFun (j : J) (x : F.obj j) : ColimitType.{w} F :=
-  Quot.mk _ (Prequotient.of j x)
-
-/-- The group homomorphism from a given abelian group in the diagram to the colimit abelian
-group. -/
-def coconeMorphism (j : J) : F.obj j ⟶ colimit.{w} F where
-  toFun := coconeFun F j
-  map_zero' := by apply Quot.sound; apply Relation.zero
-  map_add' := by intros; apply Quot.sound; apply Relation.add
+lemma Quot.map_ι [DecidableEq J] {j j' : J} {f : j ⟶ j'} (x : F.obj j) :
+    Quot.ι F j' (F.map f x) = Quot.ι F j x := by
+  dsimp [ι]
+  refine eq_of_sub_eq_zero ?_
+  erw [← (QuotientAddGroup.mk' (Relations F)).map_sub, ← AddMonoidHom.mem_ker]
+  rw [QuotientAddGroup.ker_mk']
+  simp only [DFinsupp.singleAddHom_apply]
+  exact AddSubgroup.subset_closure ⟨j, j', f, x, rfl⟩
+
+/-- (implementation detail) A morphism of commutative additive groups `Quot F →+ A`
+induces a cocone on `F` as long as the universes work out.
+-/
+@[simps]
+def toCocone [DecidableEq J] {A : Type w} [AddCommGroup A] (f : Quot F →+ A) : Cocone F where
+  pt := AddCommGrp.of A
+  ι := { app := fun j => f.comp (Quot.ι F j) }
+
+lemma Quot.desc_toCocone_desc [DecidableEq J] {A : Type w} [AddCommGroup A] (f : Quot F →+ A)
+    (hc : IsColimit c) : (hc.desc (toCocone F f)).comp (Quot.desc F c) = f := by
+  refine Quot.addMonoidHom_ext F (fun j x ↦ ?_)
+  rw [AddMonoidHom.comp_apply, ι_desc]
+  change ((c.ι.app j) ≫ hc.desc (toCocone F f)) _ = _
+  rw [hc.fac]
+  simp
+
+lemma Quot.desc_toCocone_desc_app [DecidableEq J] {A : Type w} [AddCommGroup A] (f : Quot F →+ A)
+    (hc : IsColimit c) (x : Quot F) : hc.desc (toCocone F f) (Quot.desc F c x) = f x := by
+  conv_rhs => rw [← Quot.desc_toCocone_desc F c f hc]
+  simp
 
-@[simp]
-theorem cocone_naturality {j j' : J} (f : j ⟶ j') :
-    F.map f ≫ coconeMorphism.{w} F j' = coconeMorphism F j := by
-  ext
-  apply Quot.sound
-  apply Relation.map
+/--
+If `c` is a cocone of `F` such that `Quot.desc F c` is bijective, then `c` is a colimit
+cocone of `F`.
+-/
+noncomputable def isColimit_of_bijective_desc [DecidableEq J]
+     (h : Function.Bijective (Quot.desc F c)) : IsColimit c where
+  desc s := AddCommGrp.ofHom ((Quot.desc F s).comp (AddEquiv.ofBijective
+    (Quot.desc F c) h).symm.toAddMonoidHom)
+  fac s j := by
+    ext x
+    simp only [Functor.const_obj_obj, AddEquiv.toAddMonoidHom_eq_coe, coe_comp, Function.comp_apply,
+      ofHom_apply, AddMonoidHom.coe_comp, AddMonoidHom.coe_coe]
```

Comment:
`dsimp` is usually faster that `simp` (even a "squeezed" simp), and a plain `dsimp` makes more readable code:
```suggestion
    dsimp
```